fix: rosetta not being detected when command is [rosetta]

3timeslazy opened the pull request:
Fix for the "rosetta not found" error from here: https://github.com/DataDog/orchestrion/issues/560

3timeslazy commented:
In the PR I've only added rosetta check, but if the maintainers would like, I could add qemu check as well

RomainMuller commented:
> In the PR I've only added rosetta check, but if the maintainers would like, I could add qemu check as well

That would be super helpful thank you <3

3timeslazy commented:
Added the qemu check. Also, added the `err != nil` check, because if everything ok and the err is nil, `errors.Is` will return false, while the new statement will panic. Fix that

RomainMuller commented on internal/goflags/flags.go:
```diff
@@ -399,7 +399,10 @@ func parentGoCommandFlags(ctx context.Context, pid int) (flags CommandFlags, err
 		// we try to resolve argv[1] instead. This can only manifest itself on amd64 + linux.
-		if errors.Is(err, fs.ErrNotExist) && runtime.GOARCH == "amd64" && runtime.GOOS == "linux" && filepath.Base(args[0]) == "rosetta" && len(args) > 1 {
+		notExist := (err != nil && (errors.Is(err, fs.ErrNotExist) || strings.Contains(err.Error(), "executable file not found")))
+		emulation := strings.Contains(filepath.Base(args[0]), "rosetta")
+		emulation = emulation || strings.Contains(filepath.Base(args[0]), "qemu")
```
```suggestion
		base := filepath.Base(args[0])
		emulation := strings.Contains(base, "rosetta") || strings.Contains(base, "qemu")
```

RomainMuller commented on internal/goflags/flags.go:
```diff
@@ -399,7 +399,10 @@ func parentGoCommandFlags(ctx context.Context, pid int) (flags CommandFlags, err
+		notExist := (err != nil && (errors.Is(err, fs.ErrNotExist) || strings.Contains(err.Error(), "executable file not found")))
+		emulation := strings.Contains(filepath.Base(args[0]), "rosetta")
+		emulation = emulation || strings.Contains(filepath.Base(args[0]), "qemu")
+		if notExist && runtime.GOARCH == "amd64" && runtime.GOOS == "linux" && emulation && len(args) > 1 {
```
While the Rosetta check is only relevant for AMD64, I feel the QEMU one should be relevant for more architectures?

RomainMuller replied:
That's fine by me.

RomainMuller reviewed (commented):
A couple nitpicks but nothing major. Thanks for this!
